### PR TITLE
feat: updates dfx new template for env files

### PR DESCRIPTION
**Pull request by krpeacock**
# Description

- simplifies webpack config using dotenv
- adds .env file and gitignore config
- changes the CANISTER_ID convention to match the existing one used in frontend declarations
- removes "prestart" command in package.json in favor of mentioning "npm run generate" in the README
- matches Motoko and Rust frontend instructions
- updates docs

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

**Review by lwshang — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,30 +4,37 @@
 
 ## DFX
 
+### feat: updates the dfx new starter project for env vars
+
+- Updates the starter project for env vars to use the new `dfx build` & `dfx deploy` environment variables
+- Changes the format of the canister id env vars to be `<canister_name>_CANISTER_ID`, matching the existing convention
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:10]: Matching the existing convention where?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1891,34 +1947,35 @@ You can control what will be generated and how with corresponding configuration
 
 Under dfx.json → "canisters" → "<canister_name>", developers can add a "declarations" config. Options are:
 
-* "output" → directory to place declarations for that canister | default is "src/declarations/<canister_name>"
+- "output" → directory to place declarations for that canister | default is "src/declarations/<canister_name>"
 
-* "bindings" → [] list of options, ("js", "ts", "did", "mo") | default is "js", "ts", "did"
+- "bindings" → [] list of options, ("js", "ts", "did", "mo") | default is "js", "ts", "did"
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:1952]: This mass reformatting of CHANGELOG.md is nice, but would have been better done as a separate PR, in order to make it easier to see what differences there are related to the changes in this PR.

File: src/dfx/src/lib/builders/mod.rs [hunk 3]
```diff
@@ -345,8 +345,8 @@ pub fn get_and_write_environment_variables<'a>(
     for canister in pool.get_canister_list() {
         vars.push((
             Owned(format!(
-                "CANISTER_ID_{}",
-                canister.get_name().replace('-', "_")
+                "{}_CANISTER_ID",
+                canister.get_name().replace('-', "_").to_ascii_uppercase(),
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/mod.rs:349]: This looks like a breaking change for existing projects.

Two things:
- please add a migration guide here: https://github.com/dfinity/sdk/tree/master/docs/migration
- I think we should first alter this code to add _both_ forms (if we are going to rename this), and deprecate the old form, rather than deleting it right away.

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (krpeacock) [reply CHANGELOG.md:10]: From the `process_string` we insert into `index.js.hbs`. https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Adfinity%2Fsdk.git&branch=kyle%2Frefactor-starter-for-env&file=src%2Fdfx%2Fsrc%2Flib%2Fbuilders%2Fmod.rs&editor=VSCode&version=2.2.14&start_row=276&start_col=13&end_row=276&end_col=13

File: src/dfx/src/lib/builders/mod.rs (hunk 3 above)
Comment (krpeacock) [reply src/dfx/src/lib/builders/mod.rs:349]: I don't think that we have released this yet. This is a new feature, and I noticed it wasn't compatible with the existing declarations we have been shipping, both because `CANISTER_ID` was being used as a prefix instead of suffix, and also because the canister name was lowercase.

File: CHANGELOG.md (hunk 2 above)
Comment (krpeacock) [reply CHANGELOG.md]: Reverted

File: src/dfx/assets/new_project_node_files/webpack.config.js [hunk 4]
```diff
@@ -1,41 +1,10 @@
...
-  const canisterConfig = network === "local" ? localCanisters : prodCanisters;
-
-  return Object.entries(canisterConfig).reduce((prev, current) => {
-    const [canisterName, canisterDetails] = current;
-    prev[canisterName.toUpperCase() + "_CANISTER_ID"] =
-      canisterDetails[network];
-    return prev;
-  }, {});
-}
-const canisterEnvVariables = initCanisterEnv();
```
Comment (ericswanson-dfinity) [inline src/dfx/assets/new_project_node_files/webpack.config.js:37]: Do any of these changes need to be applied to existing projects?

File: src/dfx/assets/new_project_node_files/webpack.config.js [hunk 5]
```diff
@@ -86,10 +55,10 @@ module.exports = {
       template: path.join(__dirname, frontend_entry),
       cache: false,
     }),
-    new webpack.EnvironmentPlugin({
-      NODE_ENV: "development",
-      ...canisterEnvVariables,
-    }),
+    new webpack.EnvironmentPlugin([
+      ...Object.keys(process.env).filter((key) => key.includes("CANISTER_ID")),
```
Comment (ericswanson-dfinity) [inline src/dfx/assets/new_project_node_files/webpack.config.js:59]: Are the env variables that contain `CANISTER_CANDID_PATH` in the name ever needed here?

File: docs/cli-reference/dfx-envars.md [hunk 6]
```diff
@@ -16,9 +16,9 @@ For example, if you have a `whoami_frontend` canister that lists `whoami` under
 
 ## CANISTER_ID\_{canister.name}
 
-Use environment variables with the `CANISTER_ID` prefix to reference the canister identifier for each canister in the `dfx.json` file for your project.  Hyphens are invalid in environment variables and are replaced by underscores.
+Use environment variables with the `CANISTER_ID` suffix to reference the canister identifier for each canister in the `dfx.json` file for your project. Hyphens are invalid in environment variables and are replaced by underscores.
 
-For example, if you have a `linkedup` project that consists of the `linkedup` and `connect-d` canisters, you could use the `CANISTER_ID_linkedup` and `CANISTER_ID_connect_d` environment variables to refer to the canister identifiers—for example `ryjl3-tyaaa-aaaaa-aaaba-cai` and `rrkah-fqaaa-aaaaa-aaaaq-cai`—created for your project.
+For example, if you have a `linkedup` project that consists of the `linkedup` and `connect-d` canisters, you could use the `LINKEDUP_CANISTER_ID` and `CONNECT_D_CANISTER_ID` environment variables to refer to the canister identifiers—for example `ryjl3-tyaaa-aaaaa-aaaba-cai` and `rrkah-fqaaa-aaaaa-aaaaq-cai`—created for your project.
```
Comment (ericswanson-dfinity) [inline docs/cli-reference/dfx-envars.md:21]: I guess these changes should be reverted if the new approach is
> using prefix (CANISTER_ID_<name>) for declarations going forward

File: src/dfx/assets/new_project_node_files/webpack.config.js (hunk 5 above)
Comment (krpeacock) [reply src/dfx/assets/new_project_node_files/webpack.config.js:59]: not typically

File: src/dfx/src/lib/builders/mod.rs [hunk 7]
```diff
@@ -336,17 +340,25 @@ pub fn get_and_write_environment_variables<'a>(
...
+            Owned(format!(
+                "{}_CANISTER_ID",
+                canister.get_name().replace('-', "_").to_ascii_uppercase(),
+            )),
+            Owned(canister.canister_id().to_text().into()),
+        ));
         vars.push((
             Owned(format!(
                 "CANISTER_ID_{}",
-                canister.get_name().replace('-', "_")
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/mod.rs:349]: Would you please add this form, in addition to the mix-cased form that we've had, so that this part is backwards-compatible?

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,27 @@
 
 ## DFX
 
+### feat: updates the dfx new starter project for env vars
+
+- Updates the starter project for env vars to use the new `dfx build` & `dfx deploy` environment variables
+- Changes the format of the canister id env vars to be `CANISTER_ID_<canister_name_uppercase>`, for the frontend declaraction file to be consistent with the dfx environment variables. `CANISTER_ID` as both a prefix and suffix are supported for backwards compatibility.
+
+> Migration: If your frontend project is using the old environment variables, you will need to update your `webpack.config.js` file to use the new `CANISTER_ID_prefix` variables, or to use the new webpack config provided in the starter project.
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:12]: Please add this part to a new file in https://github.com/dfinity/sdk/tree/master/docs/migration

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: src/dfx/assets/new_project_node_files/webpack.config.js (hunk 5 above)
Comment (ericswanson-dfinity) [reply src/dfx/assets/new_project_node_files/webpack.config.js:59]: That sounds like "we don't know."  In that case, please make this continue to provide the `CANISTER_CANDID_PATH` environment variables previously provided, in order to not introduce breaking changes

File: src/dfx/assets/new_project_node_files/webpack.config.js (hunk 4 above)
Comment (krpeacock) [reply src/dfx/assets/new_project_node_files/webpack.config.js:37]: They don't need to be. Existing projects only need to worry about the change to the generated file, which I noted in the migration notes. They don't need to remove this block and switch to dotenv if they don't want to

File: docs/cli-reference/dfx-envars.md [hunk 9]
```diff
@@ -16,9 +16,9 @@ For example, if you have a `whoami_frontend` canister that lists `whoami` under
 
 ## CANISTER_ID\_{canister.name}
 
-Use environment variables with the `CANISTER_ID` prefix to reference the canister identifier for each canister in the `dfx.json` file for your project.  Hyphens are invalid in environment variables and are replaced by underscores.
+Use environment variables with the `CANISTER_ID` suffix to reference the canister identifier for each canister in the `dfx.json` file for your project. Hyphens are invalid in environment variables and are replaced by underscores.
```
Comment (ericswanson-dfinity) [inline docs/cli-reference/dfx-envars.md:19]:
```suggestion
Use environment variables with the `CANISTER_ID` prefix to reference the canister identifier for each canister in the `dfx.json` file for your project. Hyphens are invalid in environment variables and are replaced by underscores.  Lowercase characters are replaced by uppercase characters.
```

File: docs/cli-reference/dfx-envars.md (hunk 6 above)
Comment (ericswanson-dfinity) [inline docs/cli-reference/dfx-envars.md:21]: I can't suggest changes, but `CANISTER_ID_LINKEDUP` and `CANISTER_ID_CONNECT_D`

File: src/dfx/assets/new_project_node_files/webpack.config.js (hunk 5 above)
Comment (krpeacock) [reply src/dfx/assets/new_project_node_files/webpack.config.js]: all current dfx environment variables will be provided now

File: CHANGELOG.md [hunk 10]
```diff
@@ -51,21 +58,24 @@ Added validate_grant_permission() and validate_revoke_permission() methods per S
...
 - `dfx nns install` had the wrong instructions for setting up the local replica type
 
 ### fix: creating an identity with `--force` no longer switches to the newly created identity
 
 ### feat(frontend-canister)!: reworked to use permissions-based access control
 
 The permissions are as follows:
-- ManagePermissions: Can grant and revoke permissions to any principal.  Controllers implicitly have this permission.
+
+- ManagePermissions: Can grant and revoke permissions to any principal. Controllers implicitly have this permission.
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:70]: Would you please submit these unrelated markdown-correctness changes as a separate PR?   It's good to clean them up, but they would be better put in a dedicated PR.

File: docs/migration/dfx-0.13.1-migration-guide.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
+# 0.13.1 Migration Guide - Environment Variables
```
Comment (ericswanson-dfinity) [inline docs/migration/dfx-0.13.1-migration-guide.md:1]: We've already released 0.13.1. Here and elsewhere:
```suggestion
# 0.13.2 Migration Guide - Environment Variables
```

File: docs/migration/dfx-0.13.1-migration-guide.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+This is not an immediate breaking change, but there will be a standardization around environment variables. Backwards compatibility will be supported until 0.15.0, but we recommend updating your projects to use the new environment variables.
+
+## Context
+
+Previously, `webpack.config.js` included an `initCanisterEnv` that parsed canister*ids.json files and mapped them to environment variables. The pattern used there was `<CANISTER_NAME_UPPERCASE>_CANISTER_ID`. This was not consistent with the pattern used in the `dfx` internals, which used `CANISTER_ID*<canister*name_case_agnostic>`. This was confusing for users, and since uppercase environment variables is a best practice, we have decided to standardize on the `CANISTER_ID*<CANISTER_NAME_UPPERCASE>` pattern across the board.
```
Comment (ericswanson-dfinity) [inline docs/migration/dfx-0.13.1-migration-guide.md:7]: Maybe I am not understanding something about the use of `*` here.

```suggestion
Previously, `webpack.config.js` included an `initCanisterEnv` that parsed canister*ids.json files and mapped them to environment variables. The pattern used there was `<CANISTER_NAME_UPPERCASE>_CANISTER_ID`. This was not consistent with the pattern used in the `dfx` internals, which used `CANISTER_ID_<canister
_name_case_agnostic>`. This was confusing for users, and since uppercase environment variables is a best practice, we have decided to standardize on the `CANISTER_ID_<CANISTER_NAME_UPPERCASE>` pattern across the board.
```

File: src/dfx/assets/new_project_motoko_files/README.md [hunk 13]
```diff
@@ -46,6 +54,7 @@ Which will start a server at `http://localhost:8080`, proxying API requests to t
 
 If you are hosting frontend code somewhere without using DFX, you may need to make one of the following adjustments to ensure your project does not fetch the root key in production:
 
-- set`NODE_ENV` to `production` if you are using Webpack
-- use your own preferred method to replace `process.env.NODE_ENV` in the autogenerated declarations
+- set`DFX_NETWORK` to `production` if you are using Webpack
```
Comment (ericswanson-dfinity) [inline src/dfx/assets/new_project_motoko_files/README.md:57]: Maybe we should reconsider this, because `DFX_NETWORK` will usually be `ic` for the production case (see index.js.hbs line 25, and builders/mod.rs line 334)

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: src/dfx/assets/new_project_motoko_files/README.md (hunk 13 above)
Comment (krpeacock) [reply src/dfx/assets/new_project_motoko_files/README.md:57]: Oh, that we can just say that. it's accurate
```suggestion
- set`DFX_NETWORK` to `ic` if you are using Webpack
```

File: docs/migration/dfx-0.13.1-migration-guide.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
+# 0.13.2 Migration Guide - Environment Variables
```
Comment (ericswanson-dfinity) [inline docs/migration/dfx-0.13.1-migration-guide.md:1]: The filename of this migration guide also needs to change

Comment (ericswanson-dfinity) [reply docs/migration/dfx-0.13.1-migration-guide.md:1]: @krpeacock Are you wanting this change to make it into dfx 0.14?  This file and its filename would need small changes.
